algorithms/rainbow/agent.py

```diff
@@ -0,0 +1,179 @@
+        self.delta_z = (args.V_max - args.V_min) / (self.atoms - 1)
+        self.batch_size = args.batch_size
+        self.n = args.multi_step
+        self.discount = args.discount
+
+        self.online_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        if args.model and os.path.isfile(args.model):
+            """
+            When you call torch.load() on a file which contains GPU tensors, those tensors will be 
+            loaded to GPU by default. You can call torch.load(.., map_location=’cpu’) and then 
+            load_state_dict() to avoid GPU RAM surge when loading a model checkpoint.
+            Source: https://pytorch.org/docs/stable/torch.html#torch.load
+            """
+            self.online_net.load_state_dict(torch.load(args.model, map_location='cpu'))
+        self.online_net.train()
+
+        self.target_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        self.update_target_net()
+        self.target_net.train()
+        for param in self.target_net.parameters():
+            param.requires_grad = False
+
+        self.optimiser = optim.Adam(self.online_net.parameters(), lr=args.lr, eps=args.adam_eps)
+
+    def reset_noise(self):
+        """ Resets noisy weights in all linear layers (of online net only) """
+        self.online_net.reset_noise()
+
+    def act(self, state):
+        """ Acts based on single state (no batch) """
+        with torch.no_grad():
+            return (self.online_net(state.unsqueeze(0)) * self.support).sum(2).argmax(1).item()
+
+    def act_e_greedy(self, state, epsilon=0.001):
+        """
+        Acts with an ε-greedy policy (used for evaluation only)
+        High ε can reduce evaluation scores drastically
+        """
+        return np.random.randint(0, self.action_space.n) if np.random.random() < epsilon \
+            else self.act(state)
+
+    def learn(self, mem):
+        """
+        Executes 1 gradient descent step sampling batch_size transitions from the memory
+        """
+        # Sample transitions
+        idxs, states, actions, returns, next_states, nonterminals, weights = \
+          mem.sample(self.batch_size)
+
+        # Calculate current state probabilities (online network noise already sampled)
+        log_ps = self.online_net(states, log=True)  # Log probabilities log p(s_t, ·; θonline)
+        log_ps_a = log_ps[range(self.batch_size), actions]  # log p(s_t, a_t; θonline)
+
+        with torch.no_grad():
+            """
+            -------------------
+            Policy Evaluation
+            -------------------
+            Calculate nth next state action probabilities with the online policy for N-step Learning
+            Probabilities: p(s_t+n, ·; θonline), i.e. for all actions
+            """
+            pns = self.online_net(next_states)
+            """ We compute the expected Q from the N-step distribution 
```

Comment:
Change 2 line comments to hashtags, makes it strange to read. 
And possibly even with 3 line comments. 
Feel free to argue

Reply:
2 line comments fine, 3 line comments no. I like more """ for long comments and it is also PEP8 so let's put that middleground

Reply:
ok for now. As long as an algorithm doesn't get broken up too much by multi-line comments i.e. over documenting